### PR TITLE
Cache on the barrelhead

### PR DESCRIPTION
**Pull request by adamemerson**
RGW Cache improvements

### CODE REVIEW

File: src/rgw/rgw_user.cc [hunk 1]
```diff
@@ -297,7 +297,7 @@ int rgw_get_user_info_from_index(RGWRados * const store,
   list<rgw_cache_entry_info *> cache_info_entries;
   cache_info_entries.push_back(&cache_info);
 
-  uinfo_cache.put(store, key, &e, cache_info_entries);
+  uinfo_cache.put(store, key, &e, { &cache_info });
```
Comment (cbodley) [inline src/rgw/rgw_user.cc]: it looks like the `cache_info_entries` list above is no longer used

**Comment by mattbenjamin** [comment]
@adamemerson promising

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by mattbenjamin** [comment]
http://pulpito.ceph.com/mbenjamin-2017-12-21_15:24:10-rgw-wip-rgw-22517---basic-smithi/

**Comment by mattbenjamin** [comment]
2017-12-21T21:49:25.384 INFO:tasks.rgw_multisite_tests:tasks.rgw_multi.tests.test_bucket_index_log_trim ... ok
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:======================================================================
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_bucket_sync_disable_enable
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-rgw-22517/qa/tasks/rgw_multi/tests.py", line 1004, in test_bucket_sync_disable_enable
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:    zonegroup_bucket_checkpoint(zonegroup_conns, bucket_name)
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-rgw-22517/qa/tasks/rgw_multi/tests.py", line 398, in zonegroup_bucket_checkpoint
2017-12-21T21:49:25.385 INFO:tasks.rgw_multisite_tests:    zone_bucket_checkpoint(target_conn.zone, source_conn.zone, bucket_name)
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-rgw-22517/qa/tasks/rgw_multi/tests.py", line 391, in zone_bucket_checkpoint
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:    (target_zone.name, source_zone.name, bucket_name)
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:AssertionError: finished bucket checkpoint for target_zone=test-zone1 source_zone=test-zone2 bucket=wyhalv-53
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:Ran 21 tests in 2924.258s
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:
2017-12-21T21:49:25.386 INFO:tasks.rgw_multisite_tests:FAILED (failures=1)
2017-12-21T21:49:25.386 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
    manager.__enter__()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/task/__init__.py", line 123, in __enter__
    self.begin()
  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-rgw-22517/qa/tasks/rgw_multisite_tests.py", line 68, in begin
    raise RuntimeError('rgw multisite test failures')
RuntimeError: rgw multisite test failures

**Comment by mattbenjamin** [comment]
@cbodley @oritwas could one of you take a look at the multisite failure (bucket sync enable-disable)--should it block this PR?

**Comment by cbodley** [comment]
> @cbodley @oritwas could one of you take a look at the multisite failure (bucket sync enable-disable)--should it block this PR?

does not look like a blocker. that same test case passed in [the other multisite job](http://pulpito.ceph.com/mbenjamin-2017-12-21_15:24:10-rgw-wip-rgw-22517---basic-smithi/1991018/)

**Comment by mattbenjamin** [comment]
@adamemerson I am getting a merge failure, can you please rebase?

**Comment by mattbenjamin** [comment]
@adamemerson never mind...

**Comment by cbodley** [comment]
@adamemerson now that this merged, i'm seeing some crashes in `ObjectCache::touch_lru()` when it calls `lru.erase(lru_iter);`. my best guess is that we're running afoul of the iterator invalidation rules (which are different between list and deque) when storing `std::deque<string>::iterator lru_iter` in `ObjectCacheEntry`

**Comment by cbodley** [comment]
added http://tracker.ceph.com/issues/22560 with stack trace

File: src/rgw/rgw_cache.cc [hunk 2]
```diff
@@ -135,8 +134,6 @@ void ObjectCache::put(string& name, ObjectCacheInfo& info, rgw_cache_entry_info
   ObjectCacheEntry& entry = iter->second;
   ObjectCacheInfo& target = entry.info;
 
-  invalidate_lru(entry);
-
```
Comment (cbodley) [inline src/rgw/rgw_cache.cc]: separately, it looks like the removal of this call is causing some failures in s3tests - added back in https://github.com/ceph/ceph/pull/19768

**Pull request merged**